Prompt API: Add overview for Prompt API (#43848)
mdn/content: files/en-us/web/api/Prompt_API/index.md

---
title: Prompt API
slug: Web/API/Prompt_API
page-type: web-api-overview
spec-urls: https://webmachinelearning.github.io/prompt-api/
---

{{DefaultAPISidebar("Prompt API")}}{{SecureContext_Header}}

> The **Prompt API** allows web pages to directly prompt a browser-provided language model through a uniform JavaScript interface, without needing to manage implementation-specific details such as prompt templating and tokenization.

Reviewer: I had to look up prompt templating. Where possible let's use terminology that high schooler could understand.

Maybe 
```suggestion
The **Prompt API** allows web pages to directly prompt a browser-provided language model through a uniform JavaScript interface, without needing to manage implementation-specific details such as how the model expects messages to be packaged and formatted.
```

Reviewer: or really simple

```suggestion
The **Prompt API** allows web pages to directly prompt a browser-provided language model through a uniform JavaScript interface, without needing to manage implementation-specific details of the AI model being used.
```

> The **Prompt API** allows web pages to directly prompt a browser-provided language model through a uniform JavaScript interface, without needing to manage implementation-specific details such as prompt templating and tokenization.

Reviewer: See my suggestions in https://github.com/mdn/content/pull/43848#discussion_r3223225349 - got applied to a wrong commit, but are still relevant.

Author: That's a link to this commit.

> The **Prompt API** allows web pages to directly prompt a browser-provided language model through a uniform JavaScript interface, without needing to manage implementation-specific details such as prompt templating and tokenization.

Reviewer: Probably also worth clarifying who provides the model and that there is only one. Specifically you can't say "I want Claude or Gemini" and you don't need to worry about that. But people might wonder.

Author: Done. 

The **Prompt API** allows web pages to directly prompt a language provided by the user agent model through a uniform JavaScript interface, without needing to manage implementation-specific details of the AI model being used. The API does not provide a way to specify a model other than the one provided by the user agent.

## Concepts and usage

### Sessions

> All interaction with the language model happens through a {{domxref("LanguageModel")}} session. A session is created by calling the static {{domxref("LanguageModel.create_static", "LanguageModel.create()")}} method, which returns a {{jsxref("Promise")}} that resolves with a `LanguageModel` instance. If the model is not yet downloaded, the browser begins the download automatically; you can monitor progress with the `monitor` option.

Reviewer: link 'monitor`

Author: I can't really link directly. I relinked create.

you can monitor progress with the `monitor` option passed to the {{domxref("LanguageModel.create_static", "LanguageModel.create()")}} method

> All interaction with the language model happens through a {{domxref("LanguageModel")}} session. A session is created by calling the static {{domxref("LanguageModel.create_static", "LanguageModel.create()")}} method, which returns a {{jsxref("Promise")}} that resolves with a `LanguageModel` instance. If the model is not yet downloaded, the browser begins the download automatically; you can monitor progress with the `monitor` option.

Reviewer: It's a little pedantic of me to ask but since we're in "concepts" it is worth stating what a session is "in this context" - usually we're after lifetime. I might assume here that a session represents all interaction with a language model for the lifetime of the page. Or perhaps it is for a single conversation.  Or perhaps it doesn't matter - this is really for any interaction and you should create a new one for every single discussion on a page that needs a new context in the model. WOuld be good to know.

Author: I've added that note to the follow up issue opened last week. #44019

Once you have a session, use {{domxref("LanguageModel.prompt()")}} to send text or multimodal input and receive the model's complete response, or {{domxref("LanguageModel.promptStreaming()")}} to receive the response incrementally as it is generated. Both methods add to the session's running context, maintaining conversational history across multiple interactions.

Use {{domxref("LanguageModel.append()")}} to preload content into the context window without generating a response — useful for providing documents, background information, or conversation history before asking a question.

Before creating a session, call the static {{domxref("LanguageModel.availability_static", "LanguageModel.availability()")}} method to determine whether the language model supports a given configuration on the current device. The method resolves with one of four string values: `"available"`, `"downloadable"`, `"downloading"`, or `"unavailable"`. This allows pages to adapt gracefully — for example, by displaying a download prompt or falling back to a server-side implementation — rather than creating a session only to have it fail.

### The context window

> Every `LanguageModel` session has a finite context window, measured in tokens, which is the total amount of input and output it can hold at once. The {{domxref("LanguageModel.contextWindow")}} property reports the session's maximum capacity, and {{domxref("LanguageModel.contextUsage")}} reports how many tokens have been consumed so far.

Reviewer: Maybe, since "amount of input and output it can hold" feels a bit odd

```suggestion
Every `LanguageModel` session has a finite context window, which constrains the total number of input and output tokens it can hold at once. The {{domxref("LanguageModel.contextWindow")}} property reports the session's maximum capacity, and {{domxref("LanguageModel.contextUsage")}} reports how many tokens have been consumed so far.
```

> When a {{domxref("LanguageModel.prompt()")}}, {{domxref("LanguageModel.promptStreaming()")}}, or {{domxref("LanguageModel.append()")}} call would exceed the context window, it throws a `QuotaExceededError` {{domxref("DOMException")}} and the {{domxref("LanguageModel.oncontextoverflow", "contextoverflow")}} event fires. To check how many tokens a piece of input would consume without actually sending it, use {{domxref("LanguageModel.measureContextUsage()")}}.

Reviewer: ```suggestion
When a {{domxref("LanguageModel.prompt()")}}, {{domxref("LanguageModel.promptStreaming()")}}, or {{domxref("LanguageModel.append()")}} call would exceed the context window, they throw a `QuotaExceededError` {{domxref("DOMException")}} and the {{domxref("LanguageModel.oncontextoverflow", "contextoverflow")}} event fires. To check how many tokens a piece of input would consume without actually sending it, use {{domxref("LanguageModel.measureContextUsage()")}}.
```

To branch from a session at a specific point in a conversation — for example, to explore different response paths in parallel without affecting each other — use {{domxref("LanguageModel.clone()")}}.

> ### Tool use

Reviewer: Just too terse - I can't guess what "tool" might mean so I determine link relevence. Maybe

```suggestion
### Trigger developer functions from prompts
```

> The Prompt API supports tool use, allowing the language model to invoke developer-defined functions during generation. Tools are registered when creating a session via the `tools` option of {{domxref("LanguageModelCreateOptions")}}. Each tool is described with a name, a natural-language description, and a JSON Schema object defining its input parameters. When the model decides to call a tool, the user agent invokes the tool's {{domxref("LanguageModelToolFunction")}} callback with the arguments the model specified, and feeds the returned string back to the model to continue generation.

Reviewer: These are now redlinks. Have you rendered this to see that it all works OK?

Author: These are dictionaries I somehow missed in my earlier pass. They are fixed.

> The Prompt API supports tool use, allowing the language model to invoke developer-defined functions during generation. Tools are registered when creating a session via the `tools` option of {{domxref("LanguageModelCreateOptions")}}. Each tool is described with a name, a natural-language description, and a JSON Schema object defining its input parameters. When the model decides to call a tool, the user agent invokes the tool's {{domxref("LanguageModelToolFunction")}} callback with the arguments the model specified, and feeds the returned string back to the model to continue generation.

Reviewer: Would be great to provide some real examples of tools. "Tool use" is a bit abstract. I would start like this.

```suggestion
The Prompt API allows the language model to invoke developer-defined functions during generation. Tools are registered when creating a session via the `tools` option of {{domxref("LanguageModelCreateOptions")}}. Each tool is described with a name, a natural-language description, and a JSON Schema object defining its input parameters. When the model decides to call a tool, the user agent invokes the tool's {{domxref("LanguageModelToolFunction")}} callback with the arguments the model specified, and feeds the returned string back to the model to continue generation.
```

### Multimodal input

Sessions can accept text, image, and audio input, depending on the capabilities of the underlying model. Declare the expected input and output modalities when creating a session using the `expectedInputs` and `expectedOutputs` options, each of which accepts an array of {{domxref("LanguageModelExpected")}} objects. These declarations also allow {{domxref("LanguageModel.availability_static", "LanguageModel.availability()")}} to check whether the desired modalities and languages are supported before committing to session creation.

> Multimodal messages are expressed using the {{domxref("LanguageModelMessage")}} and {{domxref("LanguageModelMessageContent")}} dictionaries. When a session is configured to accept images or audio, you can include `ImageBitmapSource` or `AudioBuffer` values alongside text parts in a single message.

Reviewer: The dictionaries mentioned here no longer exist. I would stucture a bit like

```suggestion
Sessions can accept text, image, and audio input, depending on the capabilities of the underlying model. 
You should test which capabilities are available before creating a session using the  {{domxref("LanguageModel.availability_static", "LanguageModel.availability()")}}  method [INSERT HERE WHY - presumably because it costs time and tokens JUST to try?].

If the model supports your desired options, you can specify them when you create a session, using the `expectedInputs` and `expectedOutputs` options,.

When a session is configured to accept images or audio, you can include `ImageBitmapSource` or `AudioBuffer` values alongside text parts in a single message.
```

---
title: Prompt API
slug: Web/API/Prompt_API
page-type: web-api-overview
spec-urls: https://webmachinelearning.github.io/prompt-api/
---

{{DefaultAPISidebar("Prompt API")}}{{SecureContext_Header}}

The **Prompt API** allows web pages to directly prompt a browser-provided language model through a uniform JavaScript interface, without needing to manage implementation-specific details such as prompt templating and tokenization.

## Concepts and usage

### Sessions

All interaction with the language model happens through a {{domxref("LanguageModel")}} session. A session is created by calling the static {{domxref("LanguageModel.create_static", "LanguageModel.create()")}} method, which returns a {{jsxref("Promise")}} that resolves with a `LanguageModel` instance. If the model is not yet downloaded, the browser begins the download automatically; you can monitor progress with the `monitor` option.

> Once you have a session, use {{domxref("LanguageModel.prompt()")}} to send text or multimodal input and receive the model's complete response, or {{domxref("LanguageModel.promptStreaming()")}} to receive the response incrementally as it is generated. Both methods add to the session's running context, maintaining conversational history across multiple turns.

Reviewer: what's a turn?

Author: "interactions" 

In next commit.

Use {{domxref("LanguageModel.append()")}} to preload content into the context window without generating a response — useful for providing documents, background information, or conversation history before asking a question.

### Checking availability

> Before creating a session, call the static {{domxref("LanguageModel.availability_static", "LanguageModel.availability()")}} method to determine whether the language model supports a given configuration on the current device. The method resolves with one of four string values: `"available"`, `"downloadable"`, `"downloading"`, or `"unavailable"`. This allows pages to adapt gracefully — for example, by displaying a download prompt or falling back to a server-side implementation — rather than creating a session only to have it fail.

Reviewer: This is probably part of the sessions heading.

You're probably wondering why I'm suggesting you remove all your stucture? Mostly because checking availability is not a concept. I might expect that in a Using the Prompt API topic.

Author: I've removed this heading.

### Tool use

The Prompt API supports tool use, allowing the language model to invoke developer-defined functions during generation. Tools are registered when creating a session via the `tools` option of {{domxref("LanguageModelCreateOptions")}}. Each tool is described with a name, a natural-language description, and a JSON Schema object defining its input parameters. When the model decides to call a tool, the user agent invokes the tool's {{domxref("LanguageModelToolFunction")}} callback with the arguments the model specified, and feeds the returned string back to the model to continue generation.

### Multimodal input

Sessions can accept text, image, and audio input, depending on the capabilities of the underlying model. Declare the expected input and output modalities when creating a session using the `expectedInputs` and `expectedOutputs` options, each of which accepts a sequence of {{domxref("LanguageModelExpected")}} objects. These declarations also allow {{domxref("LanguageModel.availability_static", "LanguageModel.availability()")}} to check whether the desired modalities and languages are supported before committing to session creation.

> Multimodal messages are expressed using the {{domxref("LanguageModelMessage")}} and {{domxref("LanguageModelMessageContent")}} dictionaries. When a session is configured to accept images or audio, you can include `ImageBitmapSource` or `AudioBuffer` values alongside text parts in a single message.

Reviewer: I think this stuff is probably all part of that first section under the main Concept heading explaining what is supported.

Author: This is a subheading of "Concepts and usage", so I'm not sure what you want. Everything will be evaluated when I overhall the page.

---
title: Prompt API
slug: Web/API/Prompt_API
page-type: web-api-overview
spec-urls: https://webmachinelearning.github.io/prompt-api/
---

{{DefaultAPISidebar("Prompt API")}}{{SecureContext_Header}}

> The **Prompt API** gives web pages the ability to directly prompt a browser-provided language model. It provides a uniform JavaScript interface that abstracts away implementation-specific details of the underlying model, such as prompt templating and tokenization. By relying on a model built into the browser, the Prompt API offers benefits including local processing of sensitive data, offline capability, model sharing across origins, and lower cost compared to cloud-based or bring-your-own-model approaches.

Reviewer 2: I would suggest shortening it to one or two sentences, as per the [template recommendation](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types/API_landing_page_template#:~:text=Begin%20the%20content%20on%20the%20page%20with%20an%20introductory%20paragraph%20%E2%80%94%20start%20by%20naming%20the%20API%20and%20saying%20what%20it%20does%2E%20This%20should%20ideally%20be%20one%20or%20two%20short%20sentences). Or splitting it into two paragraphs: the leading one, and then more details in the concepts and usage.

Something like:

```suggestion
The **Prompt API** allows web pages to directly prompt a browser-provided language model through a uniform JavaScript interface, without needing to manage implementation-specific details such as prompt templating and tokenization.
```

## Concepts and usage

### Sessions

All interaction with the language model happens through a {{domxref("LanguageModel")}} session. A session is created by calling the static {{domxref("LanguageModel.create_static", "LanguageModel.create()")}} method, which returns a {{jsxref("Promise")}} that resolves with a `LanguageModel` instance. If the model is not yet downloaded, the browser begins the download automatically; you can monitor progress with the `monitor` option.

Once you have a session, use {{domxref("LanguageModel.prompt()")}} to send text or multimodal input and receive the model's complete response, or {{domxref("LanguageModel.promptStreaming()")}} to receive the response incrementally as it is generated. Both methods add to the session's running context, maintaining conversational history across multiple turns.

> Use {{domxref("LanguageModel.append()")}} to pre-load content into the context window without generating a response — useful for injecting documents, background information, or conversation history before asking a question.

Reviewer 2: ```suggestion
Use {{domxref("LanguageModel.append()")}} to preload content into the context window without generating a response — useful for providing documents, background information, or conversation history before asking a question.
```

- Injecting has unwanted privacy connotations. I would suggest “providing”
- It’s almost never used across MDN in such form, mostly just “preload”.

### Checking availability

Before creating a session, call the static {{domxref("LanguageModel.availability_static", "LanguageModel.availability()")}} method to determine whether the language model supports a given configuration on the current device. The method resolves with one of four string values: `"available"`, `"downloadable"`, `"downloading"`, or `"unavailable"`. This allows pages to adapt gracefully — for example, by displaying a download prompt or falling back to a server-side implementation — rather than creating a session only to have it fail.

### The context window

Every `LanguageModel` session has a finite context window, measured in tokens, which is the total amount of input and output it can hold at once. The {{domxref("LanguageModel.contextWindow")}} property reports the session's maximum capacity, and {{domxref("LanguageModel.contextUsage")}} reports how many tokens have been consumed so far.

> When a `prompt()`, `promptStreaming()`, or `append()` call would exceed the context window, it throws a `QuotaExceededError` {{domxref("DOMException")}} and the {{domxref("LanguageModel.oncontextoverflow", "contextoverflow")}} event fires. To check how many tokens a piece of input would consume without actually sending it, use {{domxref("LanguageModel.measureContextUsage()")}}.

Reviewer 2: ```suggestion
When a {{domxref("LanguageModel.prompt()")}}, {{domxref("LanguageModel.promptStreaming()")}}, or {{domxref("LanguageModel.append()")}} call would exceed the context window, it throws a `QuotaExceededError` {{domxref("DOMException")}} and the {{domxref("LanguageModel.oncontextoverflow", "contextoverflow")}} event fires. To check how many tokens a piece of input would consume without actually sending it, use {{domxref("LanguageModel.measureContextUsage()")}}.
```

### Multimodal input

Sessions can accept text, image, and audio input, depending on the capabilities of the underlying model. Declare the expected input and output modalities when creating a session using the `expectedInputs` and `expectedOutputs` options, each of which accepts a sequence of {{domxref("LanguageModelExpected")}} objects. These declarations also allow {{domxref("LanguageModel.availability_static", "LanguageModel.availability()")}} to check whether the desired modalities and languages are supported before committing to session creation.

Multimodal messages are expressed using the {{domxref("LanguageModelMessage")}} and {{domxref("LanguageModelMessageContent")}} dictionaries. When a session is configured to accept images or audio, you can include `ImageBitmapSource` or `AudioBuffer` values alongside text parts in a single message.

### Permissions policy

> Access to the Prompt API is controlled by the `"language-model"` [Permissions Policy](/en-US/docs/Web/HTTP/Permissions_Policy) feature, whose default allowlist is `'self'`. This means the API is available to same-origin contexts by default. To enable it in cross-origin `<iframe>` elements, the embedding page must explicitly grant permission via the `allow` attribute or an appropriate `Permissions-Policy` response header.

Reviewer 2: ```suggestion
Access to the Prompt API is controlled by the `language-model` [Permissions Policy](/en-US/docs/Web/HTTP/Permissions_Policy) feature, whose default allowlist is `'self'`. This means the API is available to same-origin contexts by default. To enable it in cross-origin {{HTMLElement("iframe")}} elements, the embedding page must explicitly grant permission via the `allow` attribute or an appropriate `Permissions-Policy` response header.
```

  - : Options for {{domxref("LanguageModel.prompt()")}} and {{domxref("LanguageModel.promptStreaming()")}}, including an optional output constraint and an abort signal.
- {{domxref("LanguageModelAppendOptions")}}
  - : Options for {{domxref("LanguageModel.append()")}}, consisting of an abort signal.
- {{domxref("LanguageModelCloneOptions")}}
  - : Options for {{domxref("LanguageModel.clone()")}}, consisting of an abort signal.
- {{domxref("LanguageModelTool")}}
  - : Describes a tool that a language model session can invoke during generation, including its name, description, input schema, and the callback the user agent invokes on the model's behalf.
- {{domxref("LanguageModelExpected")}}
  - : Describes an expected input or output capability for a session, specifying a content type (such as `"text"`, `"image"`, or `"audio"`) and an optional list of supported languages.
- {{domxref("LanguageModelMessage")}}
  - : Represents a single message in a conversation, with a role (`"system"`, `"user"`, or `"assistant"`), content (a string or a sequence of {{domxref("LanguageModelMessageContent")}} objects), and an optional prefix flag.
- {{domxref("LanguageModelMessageContent")}}

>   - : Represents a single content part within a {{domxref("LanguageModelMessage")}}, pairing a content type with a corresponding value such as a string, `ImageBitmapSource`, or `AudioBuffer`.

Reviewer: MDN doesn't document dictionaries except in rare cases. Fold these all into the methods that use them.

As per https://github.com/mdn/content/pull/43915#pullrequestreview-4167204222
